# Cleaning Data Using pandas

# An Excel File

In [1]:
import pandas as pd  # Importing Pandas Library

In [2]:
a = "Customer Call List.xlsx"  # DataSet

In [3]:
df=pd.read_excel(a)

In [4]:
df

CustomerID First_Name    Last_Name  Phone_Number  \
0         1001      Frodo      Baggins  123-545-5421   
1         1002       Abed        Nadir  123/643/9775   
2         1003     Walter       /White    7066950392   
3         1004     Dwight      Schrute  123-543-2345   
4         1005        Jon         Snow  876|678|3469   
5         1006        Ron      Swanson  304-762-2467   
6         1007       Jeff       Winger           NaN   
7         1008   Sherlock       Holmes  876|678|3469   
8         1009    Gandalf          NaN           N/a   
9         1010      Peter       Parker  123-545-5421   
10        1011    Samwise       Gamgee           NaN   
11        1012      Harry    ...Potter    7066950392   
12        1013        Don       Draper  123-543-2345   
13        1014     Leslie        Knope  876|678|3469   
14        1015       Toby  Flenderson_  304-762-2467   
15        1016        Ron      Weasley  123-545-5421   
16        1017   Michael         Scott  123/643/9775   
17        1018      Clark         Kent    7066950392   
18        1019      Creed       Braton           N/a   
19        1020     Anakin    Skywalker  876|678|3469   
20        1020     Anakin    Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N            NaN   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes            NaN   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y            NaN   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y            NaN   
18                                    N/a             N/a            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   
20            910 Tatooine Road, Tatooine             Yes              N   

    Not_Useful_Column  
0                True  
1               False  
2                True  
3                True  
4                True  
5                True  
6               False  
7               False  
8               False  
9                True  
10               True  
11               True  
12              False  
13              False  
14              False  
15              False  
16              False  
17               True  
18               True  
19               True  
20               True

# Removing duplicate 

In [5]:
df=df.drop_duplicates()
df

CustomerID First_Name    Last_Name  Phone_Number  \
0         1001      Frodo      Baggins  123-545-5421   
1         1002       Abed        Nadir  123/643/9775   
2         1003     Walter       /White    7066950392   
3         1004     Dwight      Schrute  123-543-2345   
4         1005        Jon         Snow  876|678|3469   
5         1006        Ron      Swanson  304-762-2467   
6         1007       Jeff       Winger           NaN   
7         1008   Sherlock       Holmes  876|678|3469   
8         1009    Gandalf          NaN           N/a   
9         1010      Peter       Parker  123-545-5421   
10        1011    Samwise       Gamgee           NaN   
11        1012      Harry    ...Potter    7066950392   
12        1013        Don       Draper  123-543-2345   
13        1014     Leslie        Knope  876|678|3469   
14        1015       Toby  Flenderson_  304-762-2467   
15        1016        Ron      Weasley  123-545-5421   
16        1017   Michael         Scott  123/643/9775   
17        1018      Clark         Kent    7066950392   
18        1019      Creed       Braton           N/a   
19        1020     Anakin    Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N            NaN   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes            NaN   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y            NaN   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y            NaN   
18                                    N/a             N/a            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   

    Not_Useful_Column  
0                True  
1               False  
2                True  
3                True  
4                True  
5                True  
6               False  
7               False  
8               False  
9                True  
10               True  
11               True  
12              False  
13              False  
14              False  
15              False  
16              False  
17               True  
18               True  
19               True

# Deleting the Not-Useful-Col

In [6]:
df=df.drop(columns="Not_Useful_Column")
df

CustomerID First_Name    Last_Name  Phone_Number  \
0         1001      Frodo      Baggins  123-545-5421   
1         1002       Abed        Nadir  123/643/9775   
2         1003     Walter       /White    7066950392   
3         1004     Dwight      Schrute  123-543-2345   
4         1005        Jon         Snow  876|678|3469   
5         1006        Ron      Swanson  304-762-2467   
6         1007       Jeff       Winger           NaN   
7         1008   Sherlock       Holmes  876|678|3469   
8         1009    Gandalf          NaN           N/a   
9         1010      Peter       Parker  123-545-5421   
10        1011    Samwise       Gamgee           NaN   
11        1012      Harry    ...Potter    7066950392   
12        1013        Don       Draper  123-543-2345   
13        1014     Leslie        Knope  876|678|3469   
14        1015       Toby  Flenderson_  304-762-2467   
15        1016        Ron      Weasley  123-545-5421   
16        1017   Michael         Scott  123/643/9775   
17        1018      Clark         Kent    7066950392   
18        1019      Creed       Braton           N/a   
19        1020     Anakin    Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

## Removing the symbols like /, ..., _ etc from name col

In [7]:
df["Last_Name"] = df["Last_Name"].str.lstrip("...")
df["Last_Name"] = df["Last_Name"].str.lstrip("/")
df["Last_Name"] = df["Last_Name"].str.rstrip("_")
# or simple you also use that
# df["Last_Name"] = df["Last_Name"].str.rstrip("123._/")
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123/643/9775   
2         1003     Walter       White    7066950392   
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876|678|3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger           NaN   
7         1008   Sherlock      Holmes  876|678|3469   
8         1009    Gandalf         NaN           N/a   
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee           NaN   
11        1012      Harry      Potter    7066950392   
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876|678|3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123/643/9775   
17        1018      Clark        Kent    7066950392   
18        1019      Creed      Braton           N/a   
19        1020     Anakin   Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

In [16]:
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123/643/9775   
2         1003     Walter       White           NaN   
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876|678|3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger           NaN   
7         1008   Sherlock      Holmes  876|678|3469   
8         1009    Gandalf         NaN           N/a   
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee           NaN   
11        1012      Harry      Potter           NaN   
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876|678|3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123/643/9775   
17        1018      Clark        Kent           NaN   
18        1019      Creed      Braton           N/a   
19        1020     Anakin   Skywalker  876|678|3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

### Now cleaning the phone_numbers you see that there is a straight Integer like this 876|678|3469 having brakets nan values so we can clear it to a correct phone Number squence like this 123-545-5421 

In [27]:
df["Phone_Number"]=df["Phone_Number"].str.replace('[^a-zA-Z0-9]','', regex=True) # RegEX mean return any value 
                                                                                 # except a-zA-Z0-9 replace with nothing ''.
df["Phone_Number"]

# This modification uses regex=True to ensure that the provided pattern is treated as a regular expression. 

0     1235455421
1     1236439775
2            NaN
3     1235432345
4     8766783469
5     3047622467
6            NaN
7     8766783469
8             Na
9     1235455421
10           NaN
11           NaN
12    1235432345
13    8766783469
14    3047622467
15    1235455421
16    1236439775
17           NaN
18            Na
19    8766783469
Name: Phone_Number, dtype: object

In [28]:
df

CustomerID First_Name   Last_Name Phone_Number  \
0         1001      Frodo     Baggins   1235455421   
1         1002       Abed       Nadir   1236439775   
2         1003     Walter       White          NaN   
3         1004     Dwight     Schrute   1235432345   
4         1005        Jon        Snow   8766783469   
5         1006        Ron     Swanson   3047622467   
6         1007       Jeff      Winger          NaN   
7         1008   Sherlock      Holmes   8766783469   
8         1009    Gandalf         NaN           Na   
9         1010      Peter      Parker   1235455421   
10        1011    Samwise      Gamgee          NaN   
11        1012      Harry      Potter          NaN   
12        1013        Don      Draper   1235432345   
13        1014     Leslie       Knope   8766783469   
14        1015       Toby  Flenderson   3047622467   
15        1016        Ron     Weasley   1235455421   
16        1017   Michael        Scott   1236439775   
17        1018      Clark        Kent          NaN   
18        1019      Creed      Braton           Na   
19        1020     Anakin   Skywalker   8766783469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

## This code snippet applies a lambda function that converts each element in the "Phone_Number" column to a string. If the values were not already strings, this ensures that they are treated as such.

In [35]:
df["Phone_Number"]=df["Phone_Number"].apply(lambda x: str(x)) # coverting into string
df["Phone_Number"]

0     1235455421
1     1236439775
2            nan
3     1235432345
4     8766783469
5     3047622467
6            nan
7     8766783469
8             Na
9     1235455421
10           nan
11           nan
12    1235432345
13    8766783469
14    3047622467
15    1235455421
16    1236439775
17           nan
18            Na
19    8766783469
Name: Phone_Number, dtype: object

## If you want to add hyphens to create a specific format (e.g., xxx-xxx-xxxx), you can modify your code as follows:

### python
### df["Phone_Number"] = df["Phone_Number"].apply(lambda x: f'{x[:3]}-{x[3:6]}-{x[6:]}')
### This code uses f-strings to create the desired format by slicing the original string into three parts. It assumes that the phone number has at least 10 characters. If some entries might have fewer characters, you may need to add additional checks to handle those cases.

In [36]:
df["Phone_Number"]=df["Phone_Number"].apply(lambda x: x[0:3] + '-' + x[3:6] + '-' + x[6:10])
df["Phone_Number"]

0     123-545-5421
1     123-643-9775
2            nan--
3     123-543-2345
4     876-678-3469
5     304-762-2467
6            nan--
7     876-678-3469
8             Na--
9     123-545-5421
10           nan--
11           nan--
12    123-543-2345
13    876-678-3469
14    304-762-2467
15    123-545-5421
16    123-643-9775
17           nan--
18            Na--
19    876-678-3469
Name: Phone_Number, dtype: object

In [37]:
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White         nan--   
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger         nan--   
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf         NaN          Na--   
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee         nan--   
11        1012      Harry      Potter         nan--   
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent         nan--   
18        1019      Creed      Braton          Na--   
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

### . Now Replacing 'nan--	' by '  '

In [42]:
df["Phone_Number"]=df["Phone_Number"].str.replace('nan--', '')
df["Phone_Number"]=df["Phone_Number"].str.replace('Na--', '')
df["Phone_Number"]

0     123-545-5421
1     123-643-9775
2                 
3     123-543-2345
4     876-678-3469
5     304-762-2467
6                 
7     876-678-3469
8                 
9     123-545-5421
10                
11                
12    123-543-2345
13    876-678-3469
14    304-762-2467
15    123-545-5421
16    123-643-9775
17                
18                
19    876-678-3469
Name: Phone_Number, dtype: object

In [43]:
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White                 
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger                 
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf         NaN                 
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee                 
11        1012      Harry      Potter                 
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent                 
18        1019      Creed      Braton                 
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  
0                   123 Shire Lane, Shire             Yes             No  
1                     93 West Main Street              No            Yes  
2                      298 Drugs Driveway               N            NaN  
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y  
4                        123 Dragons Road               Y             No  
5                        768 City Parkway             Yes            Yes  
6                       1209 South Street              No             No  
7                           98 Clue Drive               N             No  
8                        123 Middle Earth             Yes            NaN  
9              25th Main Street, New York             Yes             No  
10                  612 Shire Lane, Shire             Yes             No  
11                   2394 Hogwarts Avenue               Y            NaN  
12                       2039 Main Street             Yes              N  
13                       343 City Parkway             Yes             No  
14                          214 HR Avenue               N             No  
15                   2395 Hogwarts Avenue              No              N  
16         121 Paper Avenue, Pennsylvania             Yes             No  
17                        3498 Super Lane               Y            NaN  
18                                    N/a             N/a            Yes  
19            910 Tatooine Road, Tatooine             Yes              N

In [55]:
df["Address"]

0                     123 Shire Lane, Shire
1                       93 West Main Street
2                        298 Drugs Driveway
3     980 Paper Avenue, Pennsylvania, 18503
4                          123 Dragons Road
5                          768 City Parkway
6                         1209 South Street
7                             98 Clue Drive
8                          123 Middle Earth
9                25th Main Street, New York
10                    612 Shire Lane, Shire
11                     2394 Hogwarts Avenue
12                         2039 Main Street
13                         343 City Parkway
14                            214 HR Avenue
15                     2395 Hogwarts Avenue
16           121 Paper Avenue, Pennsylvania
17                          3498 Super Lane
18                                      N/a
19              910 Tatooine Road, Tatooine
Name: Address, dtype: object

## The str.split() method is used to split a string into a list of substrings based on a specified delimiter.
## This uses the keyword argument n to specify the maximum number of splits. The resulting DataFrame will have three columns ('Street_Address', 'State', 'ZipCode') based on the split results. If there are fewer than three parts in the address, the remaining columns will contain NaN values.

In [56]:
df[['Street_Address', 'State', 'ZipCode']] = df["Address"].str.split(',', n=2, expand=True)
df[['Street_Address', 'State', 'ZipCode']]

Street_Address          State ZipCode
0         123 Shire Lane          Shire    None
1    93 West Main Street           None    None
2     298 Drugs Driveway           None    None
3       980 Paper Avenue   Pennsylvania   18503
4       123 Dragons Road           None    None
5       768 City Parkway           None    None
6      1209 South Street           None    None
7          98 Clue Drive           None    None
8       123 Middle Earth           None    None
9       25th Main Street       New York    None
10        612 Shire Lane          Shire    None
11  2394 Hogwarts Avenue           None    None
12      2039 Main Street           None    None
13      343 City Parkway           None    None
14         214 HR Avenue           None    None
15  2395 Hogwarts Avenue           None    None
16      121 Paper Avenue   Pennsylvania    None
17       3498 Super Lane           None    None
18                   N/a           None    None
19     910 Tatooine Road       Tatooine    None

In [54]:
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White                 
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger                 
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf         NaN                 
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee                 
11        1012      Harry      Potter                 
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent                 
18        1019      Creed      Braton                 
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire             Yes             No   
1                     93 West Main Street              No            Yes   
2                      298 Drugs Driveway               N            NaN   
3   980 Paper Avenue, Pennsylvania, 18503             Yes              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway             Yes            Yes   
6                       1209 South Street              No             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth             Yes            NaN   
9              25th Main Street, New York             Yes             No   
10                  612 Shire Lane, Shire             Yes             No   
11                   2394 Hogwarts Avenue               Y            NaN   
12                       2039 Main Street             Yes              N   
13                       343 City Parkway             Yes             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue              No              N   
16         121 Paper Avenue, Pennsylvania             Yes             No   
17                        3498 Super Lane               Y            NaN   
18                                    N/a             N/a            Yes   
19            910 Tatooine Road, Tatooine             Yes              N   

          Street_Address          State ZipCode  
0         123 Shire Lane          Shire    None  
1    93 West Main Street           None    None  
2     298 Drugs Driveway           None    None  
3       980 Paper Avenue   Pennsylvania   18503  
4       123 Dragons Road           None    None  
5       768 City Parkway           None    None  
6      1209 South Street           None    None  
7          98 Clue Drive           None    None  
8       123 Middle Earth           None    None  
9       25th Main Street       New York    None  
10        612 Shire Lane          Shire    None  
11  2394 Hogwarts Avenue           None    None  
12      2039 Main Street           None    None  
13      343 City Parkway           None    None  
14         214 HR Avenue           None    None  
15  2395 Hogwarts Avenue           None    None  
16      121 Paper Avenue   Pennsylvania    None  
17       3498 Super Lane           None    None  
18                   N/a           None    None  
19     910 Tatooine Road       Tatooine    None

In [57]:
df['Paying Customer']

0     Yes
1      No
2       N
3     Yes
4       Y
5     Yes
6      No
7       N
8     Yes
9     Yes
10    Yes
11      Y
12    Yes
13    Yes
14      N
15     No
16    Yes
17      Y
18    N/a
19    Yes
Name: Paying Customer, dtype: object

### We want to change the value of Yes into Y and No into No

In [58]:
df['Paying Customer']=df['Paying Customer'].str.replace('Yes','Y')
df['Paying Customer']=df['Paying Customer'].str.replace('No','N')
df['Paying Customer']

0       Y
1       N
2       N
3       Y
4       Y
5       Y
6       N
7       N
8       Y
9       Y
10      Y
11      Y
12      Y
13      Y
14      N
15      N
16      Y
17      Y
18    N/a
19      Y
Name: Paying Customer, dtype: object

In [59]:
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White                 
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger                 
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf         NaN                 
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee                 
11        1012      Harry      Potter                 
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent                 
18        1019      Creed      Braton                 
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire               Y             No   
1                     93 West Main Street               N            Yes   
2                      298 Drugs Driveway               N            NaN   
3   980 Paper Avenue, Pennsylvania, 18503               Y              Y   
4                        123 Dragons Road               Y             No   
5                        768 City Parkway               Y            Yes   
6                       1209 South Street               N             No   
7                           98 Clue Drive               N             No   
8                        123 Middle Earth               Y            NaN   
9              25th Main Street, New York               Y             No   
10                  612 Shire Lane, Shire               Y             No   
11                   2394 Hogwarts Avenue               Y            NaN   
12                       2039 Main Street               Y              N   
13                       343 City Parkway               Y             No   
14                          214 HR Avenue               N             No   
15                   2395 Hogwarts Avenue               N              N   
16         121 Paper Avenue, Pennsylvania               Y             No   
17                        3498 Super Lane               Y            NaN   
18                                    N/a             N/a            Yes   
19            910 Tatooine Road, Tatooine               Y              N   

          Street_Address          State ZipCode  
0         123 Shire Lane          Shire    None  
1    93 West Main Street           None    None  
2     298 Drugs Driveway           None    None  
3       980 Paper Avenue   Pennsylvania   18503  
4       123 Dragons Road           None    None  
5       768 City Parkway           None    None  
6      1209 South Street           None    None  
7          98 Clue Drive           None    None  
8       123 Middle Earth           None    None  
9       25th Main Street       New York    None  
10        612 Shire Lane          Shire    None  
11  2394 Hogwarts Avenue           None    None  
12      2039 Main Street           None    None  
13      343 City Parkway           None    None  
14         214 HR Avenue           None    None  
15  2395 Hogwarts Avenue           None    None  
16      121 Paper Avenue   Pennsylvania    None  
17       3498 Super Lane           None    None  
18                   N/a           None    None  
19     910 Tatooine Road       Tatooine    None

### Similarly in Do_Not_Contact also We Change Like obove

In [60]:
df["Do_Not_Contact"]

0      No
1     Yes
2     NaN
3       Y
4      No
5     Yes
6      No
7      No
8     NaN
9      No
10     No
11    NaN
12      N
13     No
14     No
15      N
16     No
17    NaN
18    Yes
19      N
Name: Do_Not_Contact, dtype: object

In [61]:
df['Do_Not_Contact']=df['Do_Not_Contact'].str.replace('Yes','Y')
df['Do_Not_Contact']=df['Do_Not_Contact'].str.replace('No','N')
df['Do_Not_Contact']

0       N
1       Y
2     NaN
3       Y
4       N
5       Y
6       N
7       N
8     NaN
9       N
10      N
11    NaN
12      N
13      N
14      N
15      N
16      N
17    NaN
18      Y
19      N
Name: Do_Not_Contact, dtype: object

In [62]:
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White                 
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger                 
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf         NaN                 
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee                 
11        1012      Harry      Potter                 
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent                 
18        1019      Creed      Braton                 
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire               Y              N   
1                     93 West Main Street               N              Y   
2                      298 Drugs Driveway               N            NaN   
3   980 Paper Avenue, Pennsylvania, 18503               Y              Y   
4                        123 Dragons Road               Y              N   
5                        768 City Parkway               Y              Y   
6                       1209 South Street               N              N   
7                           98 Clue Drive               N              N   
8                        123 Middle Earth               Y            NaN   
9              25th Main Street, New York               Y              N   
10                  612 Shire Lane, Shire               Y              N   
11                   2394 Hogwarts Avenue               Y            NaN   
12                       2039 Main Street               Y              N   
13                       343 City Parkway               Y              N   
14                          214 HR Avenue               N              N   
15                   2395 Hogwarts Avenue               N              N   
16         121 Paper Avenue, Pennsylvania               Y              N   
17                        3498 Super Lane               Y            NaN   
18                                    N/a             N/a              Y   
19            910 Tatooine Road, Tatooine               Y              N   

          Street_Address          State ZipCode  
0         123 Shire Lane          Shire    None  
1    93 West Main Street           None    None  
2     298 Drugs Driveway           None    None  
3       980 Paper Avenue   Pennsylvania   18503  
4       123 Dragons Road           None    None  
5       768 City Parkway           None    None  
6      1209 South Street           None    None  
7          98 Clue Drive           None    None  
8       123 Middle Earth           None    None  
9       25th Main Street       New York    None  
10        612 Shire Lane          Shire    None  
11  2394 Hogwarts Avenue           None    None  
12      2039 Main Street           None    None  
13      343 City Parkway           None    None  
14         214 HR Avenue           None    None  
15  2395 Hogwarts Avenue           None    None  
16      121 Paper Avenue   Pennsylvania    None  
17       3498 Super Lane           None    None  
18                   N/a           None    None  
19     910 Tatooine Road       Tatooine    None

## At Last removing the NaN, N/a and none values from our dataset

In [65]:
df=df.replace('N/a', '')
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
1         1002       Abed       Nadir  123-643-9775   
2         1003     Walter       White                 
3         1004     Dwight     Schrute  123-543-2345   
4         1005        Jon        Snow  876-678-3469   
5         1006        Ron     Swanson  304-762-2467   
6         1007       Jeff      Winger                 
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf         NaN                 
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee                 
11        1012      Harry      Potter                 
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent                 
18        1019      Creed      Braton                 
19        1020     Anakin   Skywalker  876-678-3469   

                                  Address Paying Customer Do_Not_Contact  \
0                   123 Shire Lane, Shire               Y              N   
1                     93 West Main Street               N              Y   
2                      298 Drugs Driveway               N            NaN   
3   980 Paper Avenue, Pennsylvania, 18503               Y              Y   
4                        123 Dragons Road               Y              N   
5                        768 City Parkway               Y              Y   
6                       1209 South Street               N              N   
7                           98 Clue Drive               N              N   
8                        123 Middle Earth               Y            NaN   
9              25th Main Street, New York               Y              N   
10                  612 Shire Lane, Shire               Y              N   
11                   2394 Hogwarts Avenue               Y            NaN   
12                       2039 Main Street               Y              N   
13                       343 City Parkway               Y              N   
14                          214 HR Avenue               N              N   
15                   2395 Hogwarts Avenue               N              N   
16         121 Paper Avenue, Pennsylvania               Y              N   
17                        3498 Super Lane               Y            NaN   
18                                                                     Y   
19            910 Tatooine Road, Tatooine               Y              N   

          Street_Address          State ZipCode  
0         123 Shire Lane          Shire    None  
1    93 West Main Street           None    None  
2     298 Drugs Driveway           None    None  
3       980 Paper Avenue   Pennsylvania   18503  
4       123 Dragons Road           None    None  
5       768 City Parkway           None    None  
6      1209 South Street           None    None  
7          98 Clue Drive           None    None  
8       123 Middle Earth           None    None  
9       25th Main Street       New York    None  
10        612 Shire Lane          Shire    None  
11  2394 Hogwarts Avenue           None    None  
12      2039 Main Street           None    None  
13      343 City Parkway           None    None  
14         214 HR Avenue           None    None  
15  2395 Hogwarts Avenue           None    None  
16      121 Paper Avenue   Pennsylvania    None  
17       3498 Super Lane           None    None  
18                                 None    None  
19     910 Tatooine Road       Tatooine    None

In [69]:
df=df.fillna('')
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
2         1003     Walter       White                 
4         1005        Jon        Snow  876-678-3469   
6         1007       Jeff      Winger                 
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf                             
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee                 
11        1012      Harry      Potter                 
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent                 
19        1020     Anakin   Skywalker  876-678-3469   

                           Address Paying Customer Do_Not_Contact  \
0            123 Shire Lane, Shire               Y              N   
2               298 Drugs Driveway               N                  
4                 123 Dragons Road               Y              N   
6                1209 South Street               N              N   
7                    98 Clue Drive               N              N   
8                 123 Middle Earth               Y                  
9       25th Main Street, New York               Y              N   
10           612 Shire Lane, Shire               Y              N   
11            2394 Hogwarts Avenue               Y                  
12                2039 Main Street               Y              N   
13                343 City Parkway               Y              N   
14                   214 HR Avenue               N              N   
15            2395 Hogwarts Avenue               N              N   
16  121 Paper Avenue, Pennsylvania               Y              N   
17                 3498 Super Lane               Y                  
19     910 Tatooine Road, Tatooine               Y              N   

          Street_Address          State ZipCode  
0         123 Shire Lane          Shire          
2     298 Drugs Driveway                         
4       123 Dragons Road                         
6      1209 South Street                         
7          98 Clue Drive                         
8       123 Middle Earth                         
9       25th Main Street       New York          
10        612 Shire Lane          Shire          
11  2394 Hogwarts Avenue                         
12      2039 Main Street                         
13      343 City Parkway                         
14         214 HR Avenue                         
15  2395 Hogwarts Avenue                         
16      121 Paper Avenue   Pennsylvania          
17       3498 Super Lane                         
19     910 Tatooine Road       Tatooine

### This code snippet iterates through the index of a DataFrame df and checks the value in the 'Do_Not_Contact' column for each row. If the value is 'Y' (indicating that the contact should not be made), it removes the corresponding row from the DataFrame using the drop method with inplace=True. After the loop completes, the modified DataFrame is accessible as df.

In [70]:
for x in df.index:
    if df.loc[x, 'Do_Not_Contact'] == 'Y':
        df.drop(x, inplace=True)
        
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
2         1003     Walter       White                 
4         1005        Jon        Snow  876-678-3469   
6         1007       Jeff      Winger                 
7         1008   Sherlock      Holmes  876-678-3469   
8         1009    Gandalf                             
9         1010      Peter      Parker  123-545-5421   
10        1011    Samwise      Gamgee                 
11        1012      Harry      Potter                 
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
17        1018      Clark        Kent                 
19        1020     Anakin   Skywalker  876-678-3469   

                           Address Paying Customer Do_Not_Contact  \
0            123 Shire Lane, Shire               Y              N   
2               298 Drugs Driveway               N                  
4                 123 Dragons Road               Y              N   
6                1209 South Street               N              N   
7                    98 Clue Drive               N              N   
8                 123 Middle Earth               Y                  
9       25th Main Street, New York               Y              N   
10           612 Shire Lane, Shire               Y              N   
11            2394 Hogwarts Avenue               Y                  
12                2039 Main Street               Y              N   
13                343 City Parkway               Y              N   
14                   214 HR Avenue               N              N   
15            2395 Hogwarts Avenue               N              N   
16  121 Paper Avenue, Pennsylvania               Y              N   
17                 3498 Super Lane               Y                  
19     910 Tatooine Road, Tatooine               Y              N   

          Street_Address          State ZipCode  
0         123 Shire Lane          Shire          
2     298 Drugs Driveway                         
4       123 Dragons Road                         
6      1209 South Street                         
7          98 Clue Drive                         
8       123 Middle Earth                         
9       25th Main Street       New York          
10        612 Shire Lane          Shire          
11  2394 Hogwarts Avenue                         
12      2039 Main Street                         
13      343 City Parkway                         
14         214 HR Avenue                         
15  2395 Hogwarts Avenue                         
16      121 Paper Avenue   Pennsylvania          
17       3498 Super Lane                         
19     910 Tatooine Road       Tatooine

### Similar for Phone Number

In [71]:
for x in df.index:
    if df.loc[x, 'Phone_Number'] == '':
        df.drop(x, inplace=True)
        
df

CustomerID First_Name   Last_Name  Phone_Number  \
0         1001      Frodo     Baggins  123-545-5421   
4         1005        Jon        Snow  876-678-3469   
7         1008   Sherlock      Holmes  876-678-3469   
9         1010      Peter      Parker  123-545-5421   
12        1013        Don      Draper  123-543-2345   
13        1014     Leslie       Knope  876-678-3469   
14        1015       Toby  Flenderson  304-762-2467   
15        1016        Ron     Weasley  123-545-5421   
16        1017   Michael        Scott  123-643-9775   
19        1020     Anakin   Skywalker  876-678-3469   

                           Address Paying Customer Do_Not_Contact  \
0            123 Shire Lane, Shire               Y              N   
4                 123 Dragons Road               Y              N   
7                    98 Clue Drive               N              N   
9       25th Main Street, New York               Y              N   
12                2039 Main Street               Y              N   
13                343 City Parkway               Y              N   
14                   214 HR Avenue               N              N   
15            2395 Hogwarts Avenue               N              N   
16  121 Paper Avenue, Pennsylvania               Y              N   
19     910 Tatooine Road, Tatooine               Y              N   

          Street_Address          State ZipCode  
0         123 Shire Lane          Shire          
4       123 Dragons Road                         
7          98 Clue Drive                         
9       25th Main Street       New York          
12      2039 Main Street                         
13      343 City Parkway                         
14         214 HR Avenue                         
15  2395 Hogwarts Avenue                         
16      121 Paper Avenue   Pennsylvania          
19     910 Tatooine Road       Tatooine

### At last rest the Indexes

In [73]:
df=df.reset_index(drop=True)
df

CustomerID First_Name   Last_Name  Phone_Number  \
0        1001      Frodo     Baggins  123-545-5421   
1        1005        Jon        Snow  876-678-3469   
2        1008   Sherlock      Holmes  876-678-3469   
3        1010      Peter      Parker  123-545-5421   
4        1013        Don      Draper  123-543-2345   
5        1014     Leslie       Knope  876-678-3469   
6        1015       Toby  Flenderson  304-762-2467   
7        1016        Ron     Weasley  123-545-5421   
8        1017   Michael        Scott  123-643-9775   
9        1020     Anakin   Skywalker  876-678-3469   

                          Address Paying Customer Do_Not_Contact  \
0           123 Shire Lane, Shire               Y              N   
1                123 Dragons Road               Y              N   
2                   98 Clue Drive               N              N   
3      25th Main Street, New York               Y              N   
4                2039 Main Street               Y              N   
5                343 City Parkway               Y              N   
6                   214 HR Avenue               N              N   
7            2395 Hogwarts Avenue               N              N   
8  121 Paper Avenue, Pennsylvania               Y              N   
9     910 Tatooine Road, Tatooine               Y              N   

         Street_Address          State ZipCode  
0        123 Shire Lane          Shire          
1      123 Dragons Road                         
2         98 Clue Drive                         
3      25th Main Street       New York          
4      2039 Main Street                         
5      343 City Parkway                         
6         214 HR Avenue                         
7  2395 Hogwarts Avenue                         
8      121 Paper Avenue   Pennsylvania          
9     910 Tatooine Road       Tatooine